crypto/evp/evp_extra_test.cc

```diff
@@ -898,3 +1012,98 @@ TEST(EVPExtraTest, Ed25519Keygen) {
   ASSERT_TRUE(EVP_DigestVerify(ctx.get(), sig, len,
                                reinterpret_cast<const uint8_t *>("hello"), 5));
 }
+
+struct RsassaPssParamsMatchTestInput {
+  const uint8_t *der;
+  size_t der_len;
+  const EVP_MD *signature_md;
+  const EVP_MD *rsa_mgf1_md;
+  int match;
+} kRsassaPssParamsMatchTestInputs[] = {
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha1(),
+     EVP_sha1(), 0},
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha224(),
+     EVP_sha224(), 0},
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha256(),
+     EVP_sha256(), 1},
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha384(),
+     EVP_sha384(), 0},
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha512(),
+     EVP_sha512(), 0},
+    {kExampleRSAPSSKeyPKCS8, sizeof(kExampleRSAPSSKeyPKCS8), EVP_sha512_256(),
+     EVP_sha512_256(), 0},
+};
+
+class EVPRsaPssExtraTest
+    : public testing::TestWithParam<RsassaPssParamsMatchTestInput> {};
+
+// This test checks params match.
+// All parameters in the signature structure algorithm identifier MUST
+// match the parameters in the key structure algorithm identifier except
+// the saltLength field.
+// See 3.3. https://tools.ietf.org/html/rfc4055#section-3.3
+TEST_P(EVPRsaPssExtraTest, PssParamsMatch) {
+  const auto &param = GetParam();
+  const uint8_t *p = param.der;
+  const EVP_MD *signature_md = param.signature_md;
+  const EVP_MD *rsa_mgf1_md = param.rsa_mgf1_md;
+  // Holds ownership of heap-allocated EVP_PKEY.
+  bssl::UniquePtr<EVP_PKEY> pkey_up(
+      d2i_AutoPrivateKey(nullptr, &p, param.der_len));
+  ASSERT_TRUE(pkey_up);
+  EXPECT_EQ(param.der + param.der_len, p);
+  EXPECT_EQ(EVP_PKEY_RSA_PSS, EVP_PKEY_id(pkey_up.get()));
+  // Holds ownership of heap-allocated EVP_PKEY_CTX.
+  bssl::UniquePtr<EVP_PKEY_CTX> pkey_ctx_up(
+      EVP_PKEY_CTX_new(pkey_up.get(), nullptr));
+  ASSERT_TRUE(pkey_ctx_up);
+  EVP_PKEY_CTX *pkey_ctx = pkey_ctx_up.get();
+  // Init pss params by calling |EVP_PKEY_sign_init|.
+  // These pss params are fetched from the key structure.
+  ASSERT_TRUE(EVP_PKEY_sign_init(pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_CTX_set_rsa_padding(pkey_ctx, RSA_PKCS1_PSS_PADDING));
+  // Compare one way hash func.
+  EXPECT_EQ(EVP_PKEY_CTX_set_signature_md(pkey_ctx, signature_md), param.match);
```

Comment:
I'm not sure what this is checking or comparing. Maybe I am missing something? `EVP_PKEY_CTX_set_*` typically assigns/sets a value to the public key structure.

I tried tracing the code:

`EVP_PKEY_CTX_set_signature_md` points to this function
https://github.com/awslabs/aws-lc/blob/rsa-pss/crypto/evp/evp.c#L425

Which then points to this function
https://github.com/awslabs/aws-lc/blob/rsa-pss/crypto/evp/evp_ctx.c#L191

Which, I guess, calls this callback function  (looking at the `rsa_pss_pkey_meth` structure)
https://github.com/awslabs/aws-lc/blob/rsa-pss/crypto/evp/p_rsa.c#L396
`type` is equal to `EVP_PKEY_CTRL_MD`. 

So, the piece of code being executed is the following:
```
    case EVP_PKEY_CTRL_MD:
      if (!check_padding_md(p2, rctx->pad_mode)) {
        return 0;
      }
      rctx->md = p2;
      return 1;
```
`check_padding_md` really only fails if the paading mode for the public key is set to `RSA_NO_PADDING`, but you just set it to `RSA_PKCS1_PSS_PADDING`, so that is not the case.